drivers/gpio/gpio_rts5912.c

```diff
@@ -0,0 +1,455 @@
+
+struct gpio_rts5912_config {
+	struct gpio_driver_config common;
+	volatile uint32_t *reg_base;
+	uint8_t num_pins;
+};
+
+struct gpio_rts5912_data {
+	struct gpio_driver_data common;
+	sys_slist_t callbacks;
+};
+
+static int pin_is_valid(gpio_pin_t pin, uint8_t num_pins)
+{
+	if (pin >= num_pins) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int pin_output_high(const struct device *port, gpio_pin_t pin)
+{
+	const struct gpio_rts5912_config *config = (const struct gpio_rts5912_config *)port->config;
+	volatile uint32_t *gcr = &(config->reg_base[pin]);
+
+	int err = pin_is_valid(pin, config->num_pins);
+
+	if (err) {
+		return err;
+	}
+
+	if (*gcr & GPIO_GCR_OUTMD_Msk) {
+		/* Switch I/O mode to input mode when configuration is open-drain with output high
+		 */
+		*gcr = ((*gcr & ~GPIO_GCR_DIR_Msk) | GPIO_GCR_OUTCTRL_Msk);
+	} else {
+		*gcr |= (GPIO_GCR_OUTCTRL_Msk | GPIO_GCR_DIR_Msk);
+	}
+
+	return 0;
+}
+
+static int pin_output_low(const struct device *port, gpio_pin_t pin)
+{
+	const struct gpio_rts5912_config *config = (const struct gpio_rts5912_config *)port->config;
+	volatile uint32_t *gcr = &(config->reg_base[pin]);
+
+	int err = pin_is_valid(pin, config->num_pins);
+
+	if (err) {
+		return err;
+	}
+
+	*gcr = ((*gcr & ~GPIO_GCR_OUTCTRL_Msk) | GPIO_GCR_DIR_Msk);
+
+	return 0;
+}
+
+static int gpio_rts5912_configuration(const struct device *port, gpio_pin_t pin, gpio_flags_t flags)
+{
+	const struct gpio_rts5912_config *config = (const struct gpio_rts5912_config *)port->config;
+	volatile uint32_t *gcr = &(config->reg_base[pin]);
+	uint32_t cfg_val = *gcr;
+	int err = 0;
+
+	err = pin_is_valid(pin, config->num_pins);
+	if (err) {
+		return err;
+	}
+
+	if (flags & GPIO_INPUT) {
+		cfg_val &= ~GPIO_GCR_DIR_Msk;
+		cfg_val &= ~GPIO_GCR_OUTCTRL_Msk;
+		cfg_val |= GPIO_GCR_INDETEN_Msk;
+	}
+
+	if (flags & GPIO_OPEN_DRAIN) {
+		cfg_val |= GPIO_GCR_OUTMD_Msk;
+	} else {
+		cfg_val &= ~GPIO_GCR_OUTMD_Msk;
+	}
+
+	switch (flags & (GPIO_PULL_UP | GPIO_PULL_DOWN)) {
+	case GPIO_PULL_UP:
+		cfg_val &= ~GPIO_GCR_PULLDWEN_Msk;
+		cfg_val |= GPIO_GCR_PULLUPEN_Msk;
+		break;
+	case GPIO_PULL_DOWN:
+		cfg_val &= ~GPIO_GCR_PULLUPEN_Msk;
+		cfg_val |= GPIO_GCR_PULLDWEN_Msk;
+		break;
+	default:
+		break;
+	}
+
+	switch (flags & GPIO_VOLTAGE_MASK) {
+	case GPIO_VOLTAGE_1P8:
+		cfg_val |= GPIO_GCR_INVOLMD_Msk;
+		break;
+	case GPIO_VOLTAGE_DEFAULT:
+	case GPIO_VOLTAGE_3P3:
+		cfg_val &= ~GPIO_GCR_INVOLMD_Msk;
+		break;
+	case GPIO_VOLTAGE_5P0:
+		return -ENOTSUP;
+	default:
+		break;
+	}
+
+	*gcr = cfg_val;
+
+	if (flags & GPIO_OUTPUT) {
+		if (flags & GPIO_OUTPUT_INIT_HIGH) {
+			pin_output_high(port, pin);
+		} else {
+			pin_output_low(port, pin);
+		}
+	}
+
+	return 0;
+}
+
+#ifdef CONFIG_GPIO_GET_CONFIG
+static int gpio_rts5912_get_configuration(const struct device *port, gpio_pin_t pin,
+					  gpio_flags_t *flags)
+{
+	const struct gpio_rts5912_config *config = (const struct gpio_rts5912_config *)port->config;
+	int err = 0;
```

Comment:
no need to initialize this

Reply:
We replace every **int err** to **int err = pin_is_valid(config, pin);**